### PR TITLE
Support for docker command line option --build-arg

### PR DESCRIPTION
**Pull request by lrolaz**
This pull request can fix the issue #184 and allow to set Build Args in the pom.

Here is an example:
```
<groupId>com.spotify</groupId>
<artifactId>docker-maven-plugin</artifactId>
<configuration>
    <imageName>myImage</imageName>
    <dockerDirectory>${project.basedir}/src/main/resources</dockerDirectory>
    <imageTags>
        <tag>${project.version}</tag>
    </imageTags>
    <buildArgs>
         <HTTP_PROXY>http://x.x.x.x:x</HTTP_PROXY>
         <VERSION>${project.version}</VERSION>        
    </buildArgs>
</configuration>
```

### CODE REVIEW

File: src/test/resources/dockerDirectoryArgs/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM       busybox
+MAINTAINER John Doe "user@spotify.com"
+
+ARG VERSION
+RUN echo "${VERSION}" >> /build-version
+
+RUN date >> /build-timestamp
```
Comment (davidxia) [inline src/test/resources/dockerDirectoryArgs/Dockerfile]: Do we need this line?

**Comment by davidxia** [comment]
LGTM What do you think, @mattnworb?

File: src/main/java/com/spotify/docker/BuildMojo.java [hunk 2]
```diff
@@ -779,6 +787,10 @@ public static String separatorsToUnix(final String path) {
     if (noCache) {
       buildParams.add(DockerClient.BuildParam.noCache());
     }
+    if (!buildArgs.isEmpty()) {
+      buildParams.add(DockerClient.BuildParam.create("buildargs", 
+        URLEncoder.encode(new ObjectMapper().writeValueAsString(buildArgs), "UTF-8")));
```
Comment (mattnworb) [inline src/main/java/com/spotify/docker/BuildMojo.java]: it would be better to store the ObjectMapper as a field of the class, to avoid re-instantiation.

**Comment by mattnworb** [comment]
Looks good to me besides the above comment. Thanks @lrolaz for the contribution!

I have one question though - is it possible to use maven properties in the values of the build args with this patch, or is it necessary to hook something up within the mojo that resolves the properties?

**Comment by lrolaz** [comment]
Thanks for your comments, I've juste added 2 commits to fix them.

@mattnworb, in my understanding the maven properties can be used by the plugin to set ENV variables in the Docker Image, right ? So I think it's not possible to use maven properties for ARG. My use cases are limited:
- Set the HTTP_PROXY arg to allow a dockerfile build behind a proxy.
- Set the VERSION arg from a maven property to retrieve the desired version of a package.

If you have any idea on how to use the maven properties for build args, I'll be able to implement it.

**Comment by mattnworb** [comment]
@lrolaz I was mostly just curious if properties could be used or not, I don't think it is a necessary part of adding htis.

But I did a quick test and it seems like property resolution for the build args works out of the box:

```xml
<properties>
    <foo>bar</foo>
</properties>
<build>
    <plugins>
      <plugin>
        <groupId>com.spotify</groupId>
        <artifactId>docker-maven-plugin</artifactId>
        <version>0.1-SNAPSHOT</version>
        <configuration>
          <buildArgs>
            <FOO>${foo}</FOO>
          </buildArgs>
          ...
```

In the test for this, the docker client is passed `%7B%22FOO%22%3A%22bar%22%7D` which url decodes into `{"FOO":"bar"}`

**Comment by davidmichaelkarr** [comment]
Is this not documented in the README yet?  Is there some other set of docs for the plugin that I'm missing?

**Pull request merged**